system: op-mode: T3334: allow delayed getty restart when configuring serial ports

talmakion opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
Reconfiguration of serial console settings from the serial console is not currently possible - the restart at commit terminates the login shell and does not complete the commit. 

This PR relocates the serial-getty service control portion from the conf-mode handler into an op-mode user command "restart serial-console", wrapping it in additional checks and helpful messages. The conf handler will still call the op command, which will either silently succeed by restarting units (when no active sessions) or provide instructions to complete the process. 

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->
https://vyos.dev/T3334

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
* op-mode
* system

## Proposed changes
<!--- Describe your changes in detail -->
* Relocated service control to op-mode command "restart serial-console"
* Checking for logged-in serial sessions that may be affected by getty reconfig (extracted and adapted from ticket patch)
* Warning the user when changes are committed and serial sessions are active, otherwise restart services as normal. No prompts issued during commit, all config gen/commit steps still occur except for the service restarts (everything remains consistent)
* To apply committed changes, user will need to run "restart serial-console" to complete the process or reboot the whole router

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->
* Starting with no serial ports, configuring new ones, all OK
* Adding more ports with existing commited ports, all OK
* Removing ports, working as expected
* Removing all ports, all OK
* Rebooting with working serial devices, boot-time configuration applies as expected
* Rebooting with missing serial devices (shutdown, remove VM device, power on), boot-time configuration applies as expected (missing device units created, sitting in failed state, does not affect working units)

One sample run through from the serial session itself:
```
vyos@TEST-VYOS-LEFT# set system console device ttyS1
[edit]
vyos@TEST-VYOS-LEFT# commit
There are user sessions connected via serial console that will be terminated
when serial console settings are changed.

Please ensure all settings are committed and saved before issuing a
"restart serial-console" command to apply new configuration.
[edit]
vyos@TEST-VYOS-LEFT# save
[edit]
vyos@TEST-VYOS-LEFT# exit
exit
vyos@TEST-VYOS-LEFT:~$ restart serial-console
There are user sessions connected via serial console that will be terminated
when serial console settings are changed.

Any uncommitted changes from these sessions will be lost and in-progress actions
may be left in an inconsistent state. Continue? [y/N] y

Welcome to VyOS - TEST-VYOS-LEFT ttyS0

TEST-VYOS-LEFT login: vyos
Password:
Welcome to VyOS!
[...banner snipped...]
vyos@TEST-VYOS-LEFT:~$ configure
[edit]
vyos@TEST-VYOS-LEFT# show system console
 device ttyS0 {
 }
 device ttyS1 {
 }
[edit]
vyos@TEST-VYOS-LEFT#
```

Additional run through when nobody is logged into a serial console:
```
vyos@TEST-VYOS-LEFT# delete  system console device ttyS1 
[edit]
vyos@TEST-VYOS-LEFT# commit
[edit]
vyos@TEST-VYOS-LEFT# save
[edit]
```
No interruption or messages, as expected. Service confirmed restarted. 

## Smoketest result
<!-- Provide the output of the smoketest
```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->
There does not seem to be an appropriate smoketest for this change. 

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

talmakion commented:
Apologies about the rapid fire force push followups - added a systemctl daemon-reload to the top of the op-mode command just in case, also managed to revert a previous minor cleanup edit from my private tree at the same time.

c-po commented on src/conf_mode/system_console.py:
```diff
@@ -129,12 +128,8 @@ def apply(console):
-        if os.path.exists(f'/dev/{device}'):
-            call(f'systemctl restart serial-getty@{device}.service')
+    # Service control moved to op_mode/restart_serial.py to unify checks and prompts. 
+    call('/usr/libexec/vyos/op_mode/restart_serial.py --quiet')
```
Please de-hardcode the path and refer to https://github.com/vyos/vyos-1x/blob/current/python/vyos/defaults.py#L24

c-po commented on src/op_mode/restart_serial.py:
```diff
@@ -0,0 +1,84 @@
+def get_serial_units():
+    # Since we cannot depend on the current config for decommissioned ports, 
+    # we just grab everything that systemd knows about. 
+    tmp = cmd(f'sudo systemctl list-units {GLOB_GETTY_UNITS} --all --output json --no-pager')
```
Please move the call to `sudo` outside of the op-mode script into the XML portion

c-po commented on src/op_mode/restart_serial.py:
```diff
@@ -0,0 +1,84 @@
+            cmd(f'sudo systemctl restart {unit_name}')
+        else:
+            # Deleted stubs don't need to be restarted, just shut them down.
+            cmd(f'sudo systemctl stop {unit_name}')
```
Please move the call to `sudo` outside of the op-mode script into the XML portion

c-po commented on src/op_mode/restart_serial.py:
```diff
@@ -0,0 +1,84 @@
+    for unit in units:
+        unit_name = unit['unit']
+        if os.path.exists(os.path.join(SD_UNIT_PATH, unit_name)):
+            cmd(f'sudo systemctl restart {unit_name}')
```
Please move the call to `sudo` outside of the op-mode script into the XML portion

talmakion commented:
Updated PR with requested changes.

dmbaturin commented on src/op_mode/restart_serial.py:
```diff
@@ -0,0 +1,84 @@
+
+    return connected
+
+if __name__ == '__main__':
```
Is there a reason why we cannot make it a new-style op mode script (with `vyos.opmode`)?

talmakion replied:
Only that I didn't know that was a thing, but it sounds like a good idea - I'll do some reading.

dmbaturin reviewed (changes requested):
Looks good, but I think we should use the new-style op mode here, unless there are really strong reasons not to.

talmakion commented:
Re-wrote opmode script in the new style & re-tested.

dmbaturin commented on src/conf_mode/system_console.py:
```diff
@@ -122,20 +122,17 @@ def apply(console):
 
+    # Service control moved to op_mode/restart_serial.py to unify checks and prompts. 
+    opmode_dir = directories['op_mode']
+    call(f'{opmode_dir}/serial.py restart_getty --no-prompt')
```
Should we really call it as a subprocesses (and incur the cost of interpreter startup) if both scripts are written in Python and we could import the function (either from the script, or put it somewhere in utils)?

c-po replied:
This should be put into a utility helper like `vyos.utils.serial.restart_tty()` .

We should avoid cross-calling cfg and op-mode scripts to not get into trouble with cyclic dependencies

talmakion replied:
Looks like we were both responding around the same time, I didn't see c-po's comments when writing mine. Editing my initial response. 

I'll move the bulk of things into python vyos.utils as suggested and include args to enable/disable helpful messages and user prompts.

dmbaturin commented on src/op_mode/serial.py:
```diff
@@ -0,0 +1,94 @@
+
+    return connected
+
+def restart_getty(no_prompt: bool):
```
First, I don't think we should mention "getty" in the function name. The user doesn't care what application provides the service, the logical entity that the user is working with is serial console.

Second, should we limit the user to only being able to restart _all_ consoles at once? I believe the function should provide a way to restart only specific consoles. I'm not sure what the best arguments would be: we can either make "all" the default if `console=None`, or add two mutually-exclusive arguments like `all: bool` and `console: str`. Making "all" the default is simpler and probably enough.

talmakion replied:
1. I get your point, but it's not user-facing or at least, the users are developers. It's an internal implementation with a name only the op-mode schema or direct callers needs to worry about. I changed it from something like restart_serial_console to be more precise about what was happening. It's not much effort to change it back to match the user-facing command names. 
2. I did think about this and ended up removing it from the original flat version of the op-mode script as not relevant. It used a regex pattern argument defaulting to a match-all. The issue will be how to handle interaction with the conf_mode portion - that case would be broken if the restarts/stops weren't able to consider deleted devices. Perhaps separate pattern args along the lines of restart_consoles and cleanup_consoles so it's explicit to the caller, but only restart_consoles is exposed to op-mode usage (if at all).

dmbaturin commented on src/op_mode/serial.py:
```diff
@@ -0,0 +1,94 @@
+    ports = [ x['device'] for x in units if 'device' in x ]
+    for line in cmd(f'utmpdump {UTMP_PATH}').splitlines():
+        row = line.split('] [')
+        user_name = row[3].strip()
```
I try to always include a sample output, especially when reproducing command output isn't trivial. A reader new to `utmpdump` has no idea if `row[3]` is correct or what can go wrong.

If there's an output sample, it's much easier to get what the parsing code is working with and judge if it looks right.

dmbaturin commented on op-mode-definitions/restart-serial.xml.in:
```diff
@@ -0,0 +1,13 @@
+    <children>
+      <node name="serial-console">
+        <properties>
+          <help>Restart serial console service</help>
```
I wonder if we need to think about this wording to avoid confusion with the console server from `set service console-server` (for which there is no restart command yet).

c-po replied:
What about: `Restart serial console (ttyS)`

talmakion replied:
As above - didn't see your comment when submitting mine, so editing this response. 

I was trying to fit in with the existing restart subnodes, but will work with the suggested form.

talmakion commented:
Changes made in latest revision:
* Centralised common code between completion, op-mode and conf-mode scripts in vyos.utils.serial
  * utmpdump example output with some additional brief notes included in source
* Filtering of devices is now possible in the helpers
* Selecting 1 specific device is now possible in the op-mode side with a `device` tagNode
  * Created completion script for device param

There are some more things that could be done to make this nicer, such as using the filter to only restart changed devices from conf-mode. That would be a bulk logic change in the existing conf-mode impl, which just wipes and replaces the systemd units every time it processes. I still need to put some effort into figuring out how get_config() detects changes and deletions in order to do that properly (there's plenty of examples I've seen in conf, but I haven't taken the time to pick them apart). For this PR, it would be a minor convenience (automatically skipping logged in sessions if they're not changed), so I haven't taken it that far yet. 

What's the feeling on these revisions?

dmbaturin commented on src/op_mode/serial.py:
```diff
@@ -0,0 +1,38 @@
+import vyos.opmode
+from vyos.utils.serial import restart_login_consoles as _restart_login_consoles
+
+def restart_serial_tty(device_name: typing.Optional[str]):
```
I think we should have the API match the CLI as closely as possible. This function will be automatically exposed in the GraphQL schema, so if the command is `restart serial console [device <dev>]`, the function should also be `restart_serial_console(device:  typing.Optional[str])`.

sever-sever replied:
What about “restart_console”?
Request will be RestartConsoleSerial

dmbaturin replied:
Good point, that would be even closer to the CLI.

dmbaturin reviewed (changes requested):
I think the implementation is good now, but let's make the API match the CLI format.

talmakion committed:
system: op-mode: T3334: allow delayed getty restart when configuring serial ports

* Created op-mode command "restart serial console"
* Relocated service control to vyos.utils.serial helpers, used by conf- and
  op-mode serial console handling
  * Checking for logged-in serial sessions that may be affected by getty reconfig
  * Warning the user when changes are committed and serial sessions are active,
    otherwise restart services as normal. No prompts issued during commit,
    all config gen/commit steps still occur except for the service restarts
    (everything remains consistent)
  * To apply committed changes, user will need to run "restart serial console"
    to complete the process or reboot the whole router
  * Added additional flags and target filtering for generic use of helpers.

talmakion commented:
Updated opmode func names as recommended, quickly re-tested the opmode side.

dmbaturin reviewed (approved):
Looks good to me now, thanks!

sever-sever commented:
@c-po Can you recheck the PR?

c-po committed:
system: op-mode: T3334: replace some print() statements with Warning()

Make it more obvious for the user aber the severity of his action.

c-po commented:
![image](https://github.com/user-attachments/assets/194d2b48-b80a-4a8e-b4e8-630a92ad5796)

I change a print to a Warning for the users to better indicate the severity of the action!

c-po commented:
@MergifyIo backport sagitta circinus
